### PR TITLE
Persistent execution context

### LINKED ISSUE #60: Managing context between invocations
Hi, congrats for this great project!

I do not know much about custom runtimes, sorry if this is a newbie question!

I was wondering how you manage the [AWS Lambda Execution Context](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html).

In Nodejs background processes (e.g. http, timeouts) that did not complete when the function returned are resumed in hot starts. How does it work with this runtime?

Thanks!

### LINKED ISSUE #73: Returned html has quotes around it when Content-Type is set to text/html
When you use `UseWithApiGateway` and set the `Content-Type` header to `text/html`, all returned html is wrapped in quotes like so:

image

### LINKED ISSUE #74: When using StandaloneLambda, extra quotes are put around plain text responses
Actual:

Screenshot from 2020-06-18 13-58-09
Screenshot from 2020-06-18 13-58-34

Expected:

Screenshot from 2020-06-18 14-18-45

### PR DESCRIPTION
**Pull request by dnikolovv**
The `Context` now takes a parameter that contains your custom application context. In your custom context you can contain various things such as db connections, configuration, etc.

Your custom context will be initialized once and kept alive while the lambda is warm. You can also make changes to it during your call by mutating the `IORef` that contains it.

Usage:

```haskell
data Config =
  Config
    { configConnectionPool :: ConnectionPool
    , configLogEnv         :: LogEnv
    , configCounter        :: Int }

handler :: Request -> Context Config -> IO (Either Error Response)
handler request context = do
  let configRef = customContext context
  config <- readIORef configRef
  -- You can use your config now
  runDbAction (configConnectionPool config) action
  -- You can change your context during your execution by modifying configRef
  -- On the next lambda call, the context counter will be incremented
  modifyIORef configRef (\c -> c { configCounter = (configCounter c + 1) })
```

Breaking changes:

Besides always needing to specify the additional context (of course you could always do `Context ()`), there is one more breaking change:

In your `Main.hs` file, you must have a function called `initializedContext` that is of type `IO context`. This function will be called every cold start.

```haskell
initializeContext :: IO Config
initializeContext = Server.getConfig

generateLambdaDispatcher StandaloneLambda defaultDispatcherOptions
```

Perhaps this should be a major version bump?

- [x] - Implementation
- [x] - Examples  (see [examples/wai-app](https://github.com/dnikolovv/aws-lambda-haskell-runtime/tree/persistent-execution-context/examples/wai-app))
- [x] - Update the template
- [x] - Documentation

Closes #60.
Closes #73.
Closes #74.

@NickSeagull It's definitely not perfect, but I think this is ready for review.

### CODE REVIEW

File: package.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 name: aws-lambda-haskell-runtime
-version: 2.0.6
+version: 3.0.0
```
Comment (dnikolovv) [inline package.yaml:2]: I have bumped the version to 3.0.0 as the changes are very significant. Feel free to change that to whatever you guys find suitable.

File: src/Aws/Lambda/Meta/Dispatch.hs [hunk 2]
```diff
@@ -9,32 +8,32 @@ module Aws.Lambda.Meta.Dispatch
...
-import qualified Aws.Lambda.Meta.Main as Main
-import Data.Typeable (Typeable, typeRep, Proxy (..))
+import Data.Typeable (Proxy (..), Typeable, typeRep)
 
 {-| Helper function that the dispatcher will use to
 decode the JSON that comes as an AWS Lambda event into the
 appropriate type expected by the handler.
 -}
-decodeObj :: forall a. (FromJSON a, Typeable a) => String -> Either Error.Parsing a
+decodeObj :: forall a. (FromJSON a, Typeable a) => LazyByteString.ByteString -> Either Error.Parsing a
```
Comment (dnikolovv) [inline src/Aws/Lambda/Meta/Dispatch.hs:32]: The packing and unpacking seemed redundant, so I just made it use ByteString instead.

File: doc/01-getting-started.md [hunk 3]
```diff
@@ -60,15 +60,15 @@ to the `stack.yaml` file:
 
 ```yaml
 extra-deps:
-- aws-lambda-haskell-runtime-2.0.1
+- aws-lambda-haskell-runtime-3.0.0
```
Comment (dnikolovv) [inline doc/01-getting-started.md:63]: This assumes that `aws-lambda-haskell-runtime-3.0.0` is up on Hackage.

To test it out currently, use the following

```yaml
extra-deps:
- git: https://github.com/dnikolovv/aws-lambda-haskell-runtime.git
  commit: 1cc5a39dad2b111d05e7cb6496571ad97a8cca59
```

File: src/Aws/Lambda/Runtime.hs [hunk 4]
```diff
@@ -66,17 +71,17 @@ invokeAndRun callback manager lambdaApi event context = do
 invokeWithCallback
   :: Throws Error.Invocation
   => Throws Error.EnvironmentVariableNotSet
-  => Runtime.RunCallback
+  => Runtime.RunCallback c
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime.hs:74]:
```suggestion
  => Runtime.RunCallback context
```

File: src/Aws/Lambda/Runtime.hs [hunk 5]
```diff
@@ -66,17 +71,17 @@ invokeAndRun callback manager lambdaApi event context = do
...
   -> ApiInfo.Event
-  -> Context.Context
+  -> Context.Context c
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime.hs:76]:
```suggestion
  -> Context.Context context
```

File: src/Aws/Lambda/Runtime/Publish.hs [hunk 6]
```diff
@@ -34,22 +34,22 @@ result lambdaResult lambdaApi context manager = do
...
     context
 
 -- | Publishes a runtime initialization error back to AWS Lambda
-runtimeInitError :: ToJSON err => err -> String -> Context -> Http.Manager -> IO ()
+runtimeInitError :: ToJSON err => err -> String -> Context context -> Http.Manager -> IO ()
 runtimeInitError err lambdaApi =
   publish err (Endpoints.runtimeInitError lambdaApi)
 
-publish :: ToJSON err => err -> Endpoints.Endpoint -> Context -> Http.Manager -> IO ()
+publish :: ToJSON err => err -> Endpoints.Endpoint -> Context c -> Http.Manager -> IO ()
```
Comment (NickSeagull) [inline src/Aws/Lambda/Runtime/Publish.hs:52]:
```suggestion
publish :: ToJSON err => err -> Endpoints.Endpoint -> Context context -> Http.Manager -> IO ()
```

File: src/Aws/Lambda/Runtime.hs (hunk 4 above)
Comment (dnikolovv) [reply src/Aws/Lambda/Runtime.hs:74]: Ah, I went through these ones, but somehow they've sneaked in.

**Comment by NickSeagull** [comment]
Thank you very much for this, this is sure a great improvement!

**Pull request merged**